contracts/solve/src/ERC7683/SolverNetOutbox.sol

```diff
@@ -111,80 +111,82 @@ contract SolverNetOutbox is OwnableRoles, ReentrancyGuard, Initializable, Deploy
-
-        for (uint256 i; i < prereqs.length; ++i) {
-            TokenPrereq memory prereq = prereqs[i];
-            address token = _bytes32ToAddress(prereq.token);
-            tokens[i] = token;
+    /**
+     * @notice Verifiy and execute a call. Expenses are processed and enforced.
+     * @param call  Call to execute.
+     */
+    function _executeCall(Call memory call) internal withExpenses(call.expenses) {
+        if (call.destChainId != block.chainid) revert WrongDestChain();
 
-            if (token == address(0)) {
-                if (prereq.amount >= msg.value || prereq.amount != intent.call.value) revert InvalidPrereq();
-                preBalances[i] = address(this).balance - msg.value;
-            } else {
-                preBalances[i] = token.balanceOf(address(this));
-                address spender = _bytes32ToAddress(prereq.spender);
+        address target = _bytes32ToAddress(call.target);
 
-                token.safeTransferFrom(msg.sender, address(this), prereq.amount);
-                token.safeApprove(spender, prereq.amount);
-            }
-        }
+        if (!allowedCalls[target][bytes4(call.data)]) revert CallNotAllowed();
 
-        return (tokens, preBalances);
+        (bool success,) = payable(target).call{ value: call.value }(call.data);
+        if (!success) revert CallFailed();
     }
 
-    function _executeIntent(SolverNetIntent memory intent) internal returns (uint256 nativeAmountRequired) {
-        Call memory call = intent.call;
-        address target = _bytes32ToAddress(call.target);
-
-        if (!allowedCalls[target][bytes4(call.callData)]) revert CallNotAllowed();
+    /**
+     * @notice Mark an order as filled. Require sufficient native payment, refund excess.
+     * @param orderId     ID of the order.
+     * @param srcChainId  Chain ID on which the order was opened.
+     * @param call        Call executed.
+     * @param fillHash    Hash of fill data, verifies fill matches order.
+     */
+    function _markFilled(bytes32 orderId, uint64 srcChainId, Call memory call, bytes32 fillHash) internal {
+        // mark filled on outbox (here)
+        if (_filled[fillHash]) revert AlreadyFilled();
+        _filled[fillHash] = true;
 
-        (bool success,) = payable(target).call{ value: call.value }(call.callData);
-        if (!success) revert CallFailed();
+        // mark filled on inbox
+        uint256 fee = xcall({
+            destChainId: srcChainId,
+            conf: ConfLevel.Finalized,
+            to: _inbox,
+            data: abi.encodeCall(ISolverNetInbox.markFilled, (orderId, fillHash)),
+            gasLimit: MARK_FILLED_GAS_LIMIT
+        });
+        if (msg.value - call.value < fee) revert InsufficientFee();
+
+        // refund any overpayment in native currency
+        uint256 refund = msg.value - call.value - fee;
+        if (refund > 0) msg.sender.safeTransferETH(refund);
 
-        return call.value;
+        emit Filled(orderId, fillHash, msg.sender);
```

Comment:
kinda co-mingled marking fulfillment w/ enforcing native payment here. could be improve. might be worth just not using a help for this and keeping it flat in `fill()`